File: modules/reservation/src/main/java/com/media/bus/reservation/client/StopServiceClient.java
package com.media.bus.reservation.client;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.media.bus.common.security.TokenProvider;
import com.media.bus.reservation.dto.response.StopInfo;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

import java.util.List;
import java.util.UUID;

/**
 * stop 서비스 내부 API 클라이언트.
 *
 * 설계 의도:
 * - 유저 요청 컨텍스트와 무관하게 항상 S2S 토큰을 사용합니다.
 *   (internalRestClient는 유저 컨텍스트가 있으면 유저 JWT를 전파하므로 이 용도에 부적합)
 * - X-Service-Token 헤더에 S2S JWT를 주입 → stop의 S2STokenFilter 통과
 * - 호출 시마다 토큰을 새로 생성하여 만료 위험을 최소화합니다.
 */
@Slf4j
@Component
public class StopServiceClient {

    private static final String S2S_TOKEN_HEADER = "X-Service-Token";
    private static final String INTERNAL_STOP_PATH = "/api/v1/internal/stop";

    private final RestClient restClient;
    private final TokenProvider tokenProvider;

    public StopServiceClient(
            @Value("${services.stop.url}") String stopServiceUrl,
            TokenProvider tokenProvider
    ) {
        this.tokenProvider = tokenProvider;
        // stop 내부 API 전용 RestClient — 인증 헤더 자동 전파 없이 순수 HTTP 클라이언트로 사용
        this.restClient = RestClient.builder()
                .baseUrl(stopServiceUrl)
                .build();
    }

    /**
     * pk(UUID) 기준으로 정류소를 조회합니다.
     * 존재하지 않으면 빈 리스트를 반환합니다.
     */
    public List<StopInfo> getStopByPk(UUID pk) {
        log.debug("[StopServiceClient] pk 기준 정류소 조회: pk={}", pk);
        return fetch("pk", pk.toString());
    }

    /**
     * stopId(정류소 번호) 기준으로 정류소를 조회합니다.
     */
    public List<StopInfo> getStopByStopId(String stopId) {
        log.debug("[StopServiceClient] stopId 기준 정류소 조회: stopId={}", stopId);
        return fetch("stopId", stopId);
    }

    private List<StopInfo> fetch(String paramName, String paramValue) {
        StopPageResponse response = restClient.get()
                .uri(uriBuilder -> uriBuilder
                        .path(INTERNAL_STOP_PATH)
                        .queryParam(paramName, paramValue)
                        .build())
                .header(S2S_TOKEN_HEADER, tokenProvider.generateS2SToken())
                .retrieve()
                .body(StopPageResponse.class);

        if (response == null || response.data() == null) {
            return List.of();
        }
        return response.data().list() != null ? response.data().list() : List.of();
    }

    /**
     * stop 내부 API PageView 응답 역직렬화용 내부 레코드
     * stop 모듈 클래스를 직접 참조하지 않기 위해 클라이언트 내부에 선언
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    private record StopPageResponse(StopPageData data) {

        @JsonIgnoreProperties(ignoreUnknown = true)
        private record StopPageData(List<StopInfo> list) {}
    }
}

File: modules/iam/src/main/kotlin/com/media/bus/iam/client/stop/StopServiceClient.kt
package com.media.bus.iam.client.stop

import com.media.bus.iam.client.stop.dto.*
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.math.BigDecimal
import java.util.*

/**
 * ## stop 서비스 내부 API 클라이언트
 *
 * 설계 의도:
 * - `StopApi`(`@HttpExchange` 프록시)에 HTTP 호출을 위임하고,
 *   응답 변환(`extractList`) 등 비즈니스 로직을 이 클래스에서 처리한다.
 * - 소비자의 공개 API는 변경하지 않는다.
 */
@Component
class StopServiceClient(
    private val stopApi: StopApi,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    /** pk(UUID) 기준으로 정류소를 조회한다. 존재하지 않으면 빈 리스트를 반환한다. */
    fun getStopByPk(pk: UUID): List<StopInfo> {
        log.debug("[StopServiceClient] pk 기준 정류소 조회: pk={}", pk)
        return extractList(stopApi.getStopByPk(pk))
    }

    /** stopId(정류소 번호) 기준으로 정류소를 조회한다. 존재하지 않으면 빈 리스트를 반환한다. */
    fun getStopByStopId(stopId: String): List<StopInfo> {
        log.debug("[StopServiceClient] stopId 기준 정류소 조회: stopId={}", stopId)
        return extractList(stopApi.getStopByStopId(stopId))
    }

    /** stopName(이름 완전 일치) 기준으로 정류소를 조회한다. 동명 정류소 시 다건 반환. 없으면 빈 리스트. */
    fun getStopByStopName(stopName: String): List<StopInfo> {
        log.debug("[StopServiceClient] stopName 기준 정류소 조회: stopName={}", stopName)
        return extractList(stopApi.getStopByStopName(stopName))
    }

    /**
     * pk(UUID) 복수 기반 일괄 조회. 누락된 id는 응답 리스트에 포함되지 않는다.
     * 빈 입력에 대해서는 S2S 호출 없이 빈 리스트를 반환한다.
     */
    fun getStopsByPks(pks: Collection<UUID>): List<StopInfo> {
        if (pks.isEmpty()) return emptyList()
        log.debug("[StopServiceClient] pk 기준 일괄 정류소 조회: count={}", pks.size)
        return extractList(stopApi.getStopsByPks(BulkStopLookupRequest(ids = pks.toList())))
    }

    private fun extractList(response: StopPageResponse?): List<StopInfo> =
        response?.data?.list ?: emptyList()

    /** 정류소 단가를 조회한다. 등록된 단가가 없으면 null을 반환한다. */
    fun getStopPrice(stopId: UUID): StopPriceInfo? {
        log.debug("[StopServiceClient] 정류소 단가 조회: stopId={}", stopId)
        return stopApi.getStopPrice(stopId)?.data
    }

    /** 정류소 단가를 신규 등록한다. */
    fun createStopPrice(stopId: UUID, unitPrice: BigDecimal, registeredById: UUID): StopPriceInfo {
        log.debug("[StopServiceClient] 정류소 단가 등록: stopId={}", stopId)
        val response = stopApi.createStopPrice(CreateStopPriceRequest(stopId, unitPrice, registeredById))
        return response?.data ?: throw IllegalStateException("정류소 단가 등록에 실패했습니다.")
    }

    /** 정류소 단가를 수정한다. */
    fun updateStopPrice(stopId: UUID, unitPrice: BigDecimal, changedById: UUID): StopPriceInfo {
        log.debug("[StopServiceClient] 정류소 단가 수정: stopId={}", stopId)
        val response = stopApi.updateStopPrice(stopId, UpdateStopPriceRequest(unitPrice, changedById))
        return response?.data ?: throw IllegalStateException("정류소 단가 수정에 실패했습니다.")
    }

    /** 정류소 단가를 삭제한다. */
    fun deleteStopPrice(stopId: UUID, changedById: UUID) {
        log.debug("[StopServiceClient] 정류소 단가 삭제: stopId={}", stopId)
        stopApi.deleteStopPrice(stopId, changedById)
    }
}
